lib/segment/src/entry/entry_point.rs

```diff
@@ -297,13 +306,35 @@ pub trait SegmentEntry: SnapshotEntry {
         field_schema: Option<&PayloadFieldSchema>,
         hw_counter: &HardwareCounterCell,
     ) -> OperationResult<bool> {
-        let Some((schema, index)) =
-            self.build_field_index(op_num, key, field_schema, hw_counter)?
-        else {
-            return Ok(false);
+        let Some(field_schema) = field_schema else {
+            // Legacy case, where we tried to automatically detect the schema for the field.
+            // We don't do this anymore, as it is not reliable.
+            return Err(OperationError::TypeInferenceError {
+                field_name: key.clone(),
+            });
         };
 
-        self.apply_field_index(op_num, key.to_owned(), schema, index)
+        self.delete_field_index_if_incompatible(op_num, key, field_schema)?;
+
+        let (schema, indexes) =
+            match self.build_field_index(op_num, key, field_schema, hw_counter)? {
+                BuildFieldIndexResult::SkippedByVersion => {
+                    return Ok(false);
+                }
+                BuildFieldIndexResult::AlreadyExists => {
+                    return Ok(false);
+                }
+                BuildFieldIndexResult::IncompatibleSchema => {
+                    // This is a service error, as we should have just removed the old index
+                    // So it should not be possible to get this error
+                    return Err(OperationError::service_error(format!(
+                        "Incompatible schema for field index on field {key}",
+                    )));
+                }
```

Comment:
This handling only makes sense since we called `delete_field_index_if_incompatible` before